### ISSUE TITLE
Issues with "software version consistency" feature

### ISSUE THREAD

**Issue by gilad-yadgar** [aws/containers-roadmap#2394]
EDIT: this is related to the "software version consistency" feature launch, see What's New post: https://aws.amazon.com/about-aws/whats-new/2024/07/amazon-ecs-software-version-consistency-containerized-applications/

### Summary
since our EC2 upgraded to ecs-agent v1.83.0, images used for containers are with sha digest and not image tag

### Description
 we started getting different image value for the '{{.Config.Image}}' property using docker inspect in our ECS EC2. 
we are getting sha digest as the .Config.Image instead of getting the image tag.
the task definition contains the correct image tag (and not the digest)

we need the image tag since we rely on that custom tag to understand what was deployed. what can be done?

### Expected Behavior
we expect to see the image tag used for the container

### Observed Behavior
we get image digest used for the container

### Environment Details

* curl http://localhost:51678/v1/metadata
{"Cluster":"xxxxr","ContainerInstanceArn":"xxx","Version":"Amazon ECS Agent - v1.83.0 (*xxx"}

**Comment by sparrc** [comment]
Hi all, I have transferred this issue into the containers-roadmap repo. As far as I understand it, people are experiencing issues with this feature as a whole, rather than an issue with the ECS agent behavior specifically. For reference, see what's new post: https://aws.amazon.com/about-aws/whats-new/2024/07/amazon-ecs-software-version-consistency-containerized-applications/

Please feel free to continue adding your +1 and providing feedback :)

**Comment by simonulbrich** [comment]
This issue is affecting us as well. we utilize an initialization container that runs before the app container. this init container sets up monitoring integrations and settings that are not critical to the app itself but with a limited team we rely on the mutable tags to handle the "rolling" update as tasks are restarted. to force an application deployment for each application that my team manages for these small config updates would be an impossible task. Is there any way at all to prevent this "consistency" feature for a single container or disable entirely at the task level? 

It seems like this feature was already solved with tag immutability, giving us the option to use mutable tags if we actually needed that behavior.

**Comment by vat-gatepost-BARQUE** [comment]
We are also having an issue with this. Our development environment is setup to have all the services on a certain tag that keeps us from having to redeploy. They can simply stop the service and it comes back up with the most current image with that tag. Now they are having to update the service which is more steps than needed. This also seems to be a problem with our lambdas that spin up Fargate tasks and those tasks are not the most current version of the tag now. The update service is not an option on these so we are still trying to work that out.

**Comment by mvanholsteijn** [comment]
The strangest thing is that the feature was already available for those who wanted this. You can specify the container image with digest and that would pin the image explicitly. No code changes required to ECS.

floating potentially inconsistent -> `my-cool-image:latest`
fixed ->  `my-cool-image:latest@sha256:fdcfbed7e0a7beb8738e00fe8961c8e33e17bdeee94eab52cb8b85de1d04d024 `

**Comment by tomkins** [comment]
Also had an issue with one of our sites which I believe is related to this - a container pulling from an ECR repository with a lifecycle policy, EC2 instance restarts - and ECS wants to pull the non-existant old image as there hasn't been a fresh deploy of a container for weeks.

The version consistency is a fantastic feature, but there are situations where I **want** the tag to be used rather than the image digest at last deploy.

**Comment by vibhav-ag** [comment]
Sorry for the late response on this thread- we're aware of the impact this change has had and apologize for the churn this rollout has created. We've been actively working through the set of issues that have been highlighted on this thread and have 2 updates to share: 1/for customers who've been impacted by the lack of ability to see image tag information, we're working on a change that will bring back image tag information in the describe-tasks response, in the same format as was available prior to the release of version consistency (i.e `image:tag`). An important thing to keep in mind here is that if you run docker ps on the host, you will see the image in format `image:tag` but docker inspect will return `image:tag@digest`. 2/ We're also working on adding a configuration in the container definition that will allow you to opt-out of digest resolution for specific containers within your task- this should address both customers who want to completely opt out of digest resolution as well as customers who want to disable resolution for specific sidecar containers. I'll be using this issue to share updates on the change to bring back image tag information in describe-tasks and use issue [#2393](https://github.com/aws/containers-roadmap/issues/2393) for the change to disable digest resolution for specific containers. We're tracking both changes at high-priority.
 
Once again, we regret the churn this change has caused you all. While we still believe version consistency is the right behavior for the majority of applications hosted on ECS, we fully acknowledge that we could have done a better job socializing these changes and addressing these issues before, rather than after making the change.

**Comment by nitrotm** [comment]
I can concur that this "software version consistency" change to ECS render the concept of services totally useless for us. We may have to fallback to manually deployed task (without services) but then we'll loose the watchdog aspects which we have to re-implement ourselves.

In short, we need to guarantee a few properties on our services running background jobs:

1. a task cannot be stopped automatically within a deterministic time-frame. Therefore, we internally flag the task to stop accepting new jobs and letting it complete its currently assigned job. Only then, when the task is idle, we stop it and we relied on a nice property of ECS that it would automatically fetch that last container image associated with the tag (eg. 'latest').
2. some of our services need to have at least N tasks up and running at all time (guaranteeing some kind of always-on property)
3. some services are dynamically adjusted via auto-scaling groups due to the highly changing nature of the demand

These combined with the new constraint that all of the tasks within a service need to have the same image digest, means that we cannot roll out any update to our containers without breaking at least one property.

Tbh this feels like we may want to switch to a plain k8s solution were we can setup and manage our workloads with some degree of flexibility.

**Comment by peterjroberts** [comment]
The forced addition of this feature also caused a significant production incident for us. We deliberately used mutable tags as part of our deployments, and an ECR lifecycle policy to remove the old untagged images after a period.

This should absolutely have been an opt-in feature, or opt-out but disabled for existing services. I'm glad to see that's now been identified and raised but should this feature not be reverted until that option is available? To prevent everyone affected having to redesign workflows or implement workarounds.

As has been pointed out by others, those that want consistency by container digest can already achieve that through either tag immutability, or referring to the digest explicitly in the task definition.

**Comment by ollypom** [comment]
A quick update on @vibhav-ag's [post](https://github.com/aws/containers-roadmap/issues/2394#issuecomment-2274029246). We have now completed the first action in his comment.

Amazon ECS no longer modifies the `containerImage` field found in the [DescribeTask](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_DescribeTasks.html) or [Task Metadata](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/fargate-metadata.html) output. As part of the initial ECS software version consistency release the container image tag `imageUri:tag` was replaced with the container image digest `imageUri@digest`, this is no longer the case.

**Comment by danbf** [comment]
this almost knocked down our production environment, it did knock down stage, because we had been treating our ECR images as `mutable` and since we were running some services on `FarGate` our task instances were rotated in the background.  this prevented those `FarGate` hosted services from re-provisioning new task instances.

turns out to be not hard for us to switch to only `immutable` ECR images, and we had been taking all `public` images and enforcing in our services to only pull from a pinned copy in our ECR repos, so we were not exposed to the issues with using a `latest` tag, which we basically don't even do internally.

but wow, this breaking change hit us out of left field, and it should probably have been listed as `Breaking` on the changelog.

**Comment by gregtws** [comment]
This has impacted us as well.  We use the equivalent of a mutable 'latest' tag and perform rolling service upgrades when we move the 'latest' tag.   This lets us slowly do blue/green deployments (as our service can be told to recycle itself over time).

Instead we weren't actually progressing our blue/green deployment as AWS kept deploying the old revision of the service rather than the one pointed to by our mutable tag.  Even replacing the EC2 instance didn't fix it.  Only re-running the service deployment has done the trick.    This is a massive behavior change and shouldn't have ever been released without opting into the change.

**Comment by rs-garrick** [comment]
Just to add another use-case...

I have an app with very long-running background processes.  This app is not deployed with `--force-new-deployment` because of other issues with ECS deploys related to long-running processes that take days to exit. 

Instead, all instances are marked DRAINING so that new instances are created with the updated container image. Because the service revision is never updated with the new sha, the new instances pull down an old container image.

Oddly, there's no way to update the service revision with the new sha without triggering an actual deploy. This is the missing piece for me.

I need a way to update the sha stored in the service revision without triggering a deployment. Something like `aws ecs update-service --new-service-revision` or `--update-container-sha` (instead of `--force-new-deployment`) would be fine.

```
$ aws ecs describe-service-revisions --service-revision-arns arn:aws:ecs...
...
            "containerImages": [
                {
                    "containerName": "foo",
                    "imageDigest": "sha256:483...76d",   <-- let me update this!!
                    "image": "3622.../foo:prd"
                }
            ],
```

**Comment by vibhav-ag** [comment]
Update #2: you now have the ability to disable consistency for specific containers in your task by configuring the new versionConsistency field for each container in the task definition. Any changes to this property are applied after a deployment. Once again, we regret the churn this change has caused you all. 

[What’s New Post](https://aws.amazon.com/about-aws/whats-new/2024/11/amazon-ecs-configure-software-version-consistency/)

**Comment by vibhav-ag** [comment]
> > 2/ We're also working on adding a configuration in the container definition that will allow you to opt-out of digest resolution for specific containers within your task- this should address both customers who want to completely opt out of digest resolution as well as customers who want to disable resolution for specific sidecar containers.
> 
> If all containers in the task are opted out, will this remove the latency impact associated with this feature as well? I have a service that's updated with great frequency and whose deployments are latency sensitive. Two of the three containers in this service are already deployed with digests, but the third uses a tag because its image is built/published by CDK, and it's not possible to get access to the digest of such images to use in task definitions. So I believe we are stuck with the latency impact of this feature for mostly no reason: the image tags produced by CDK already approximate the behavior of digests, in that the image should not change if the tag is not changing.
> 
> I'm specifically referring to this line [from the documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/deployment-type-ecs.html#deployment-container-image-stability):
> 
> > To avoid potential latency altogether, specify container image digests in your task definition.

Yes, if you opt out every container, you will see no impact to deployment latency because of digest resolution.

**Comment by kaisewhite** [comment]
This causes major issues because if the sha no longer exists ecs still tries pull it. This is horrible when you are trying to deploy and test changes frequently.

**Comment by abeowlu** [comment]
@amogh09 there is an ~~un~~expected beahviour the software version veracity feature introduces in certain scenarios as well;
- since it starts 1 task of the task set to establish the image info before starting other tasks in the ecs-deployment
- in scenario where scenarios where 1 task takes a long time to start (almost 20 min), all the other tasks are simply not started
---
- cant recall if this has always been the case for ecs-deployment rollout. But it impacts the speed of rolling out quick fix, where 1 task (EC2 cluster) can potentially hold up other tasks from being created at all.
